loki/transformations/loop_blocking.py

```diff
@@ -0,0 +1,121 @@
+
+    outer_bounds = LoopRange((IntLiteral(1), num_blocks_var))
+    outer_body = (
+        Assignment(block_start,
+                   Sum(children=(
+                       Product(children=(
+                           Sum(children=(
+                               outer_loop_var, IntLiteral(-1)
+                           )),
+                           block_size_var
+                       )),
+                       IntLiteral(1)
+                   ))
+                   ),
+        Assignment(block_end,
+                   InlineCall(DeferredTypeSymbol('MIN', scope=routine),
+                              parameters=(Product(children=(outer_loop_var, block_size_var)),
+                                          loop.bounds.upper))
+                   ))
+    outer_loop = Loop(variable=outer_loop_var, body=outer_body + (inner_loop,),
+                      bounds=outer_bounds)
+
+    change_map[loop].append(outer_loop)
+
+    Transformer(change_map, inplace=True).visit(routine.ir)
+    return loop.variable, inner_loop, outer_loop
+
+
+def blocked_shape(a: Array, blocking_var: Variable, loop: ir.Loop):
+    dims = tuple(RangeIndex(
+        (1, loop.bounds.upper)) if isinstance(dim, Scalar) and blocking_var in dim else dim for dim
+                 in a.dimensions)
+    return dims
+
+
+def blocked_type(a: Array):
+    return a.type.clone(intent=None)
+
+
+def block_loop_dims(a: Array, blocking_var: Variable, loop: ir.Loop):
+    dims = tuple(
+        loop.variable if isinstance(dim, Scalar) and blocking_var in dim else dim for dim
+        in a.dimensions)
+    return dims
+
+
+def insert_block_loop_arrays(routine: Subroutine, loop_var: Variable, inner_loop: ir.Loop,
+                             outer_loop: ir.Loop, block_size: int):
+    """
+    This routine declares array variables to hold the blocks of the arrays used inside
+    the loop and replaces array variables inside the loop with their blocked counterparts
+    """
+    # Declare Blocked arrays
+    arrays = [var for var in FindVariables().visit(inner_loop.body) if
+              isinstance(var, Array) and loop_var in var]
+    name_map = {a.name: a.name + '_block' for a in arrays}
+    block_arrays = tuple(
+        a.clone(name=name_map[a.name], dimensions=blocked_shape(a, loop_var, inner_loop),
+                type=blocked_type(a)) for a in arrays)
+    routine.variables += block_arrays
+
+    block_array_expr = [
+        a.clone(name=name_map[a.name], dimensions=block_loop_dims(a, loop_var, inner_loop)) for a in
+        arrays]
+    # Replace arrays in loop with blocked arrays and update idx
+    body = SubstituteExpressions(dict(zip(arrays, block_array_expr))).visit(inner_loop.body)
+    inner_loop._update(body=body)
+
+
+def handle_device_transfers(routine: Subroutine, loop_var: Variable, inner_loop: ir.Loop,
+                            outer_loop: ir.Loop, block_size: int):
+    pass
+
+
+def block_loop(routine: Subroutine, loop: ir.Loop, block_size: int):
```

Comment:
This, as well as the other routines, would benefit from a docstring. Since `block_loop` is going to be the outward-facing API, it would be great to have even a little code example to illustrate what is being done - see for example:
https://github.com/ecmwf-ifs/loki/blob/9a76c212914780570585e5ac34bc52ef69cb60f5/loki/transformations/block_index_transformations.py#L23-L71